File: apps/web/lib/analytics.ts
import type {
  BacktestEvaluation,
  ConfidenceCalibrationBucket,
  DailyPerformance,
  HistoricalBatch,
  HistoricalRecommendationPlan,
  HorizonId,
  OutcomeResult,
  RecommendationDataset,
  StockPerformanceHistoryEntry
} from "@/lib/types";

const CLOSED_RESULTS: OutcomeResult[] = ["target_hit", "stop_loss_hit"];
const HORIZON_LABELS: Record<HorizonId, string> = {
  single_day: "Single-day",
  swing: "Swing",
  position: "Position",
  long_term: "Long-term"
};
const HORIZON_THRESHOLDS: Record<HorizonId, number> = {
  single_day: 54,
  swing: 56,
  position: 58,
  long_term: 60
};

function roundMetric(value: number) {
  return Math.round(value * 100) / 100;
}

function isClosedResult(result: OutcomeResult) {
  return CLOSED_RESULTS.includes(result);
}

function isSuccessfulResult(result: OutcomeResult) {
  return result === "target_hit";
}

function isRecommendedPlan(plan: { isRecommended?: boolean }) {
  return plan.isRecommended ?? true;
}

function planAverageReturn(plans: HistoricalRecommendationPlan[]) {
  const closedPlans = plans.filter(
    (plan) => isRecommendedPlan(plan) && isClosedResult(plan.outcome.result)
  );

  if (!closedPlans.length) {
    return null;
  }

  const average =
    closedPlans.reduce((total, plan) => total + plan.outcome.returnPct, 0) / closedPlans.length;

  return roundMetric(average);
}

function average(values: Array<number | null | undefined>) {
  const resolved = values.filter((value): value is number => value !== null && value !== undefined);

  if (!resolved.length) {
    return null;
  }

  return roundMetric(resolved.reduce((total, value) => total + value, 0) / resolved.length);
}

function maxDrawdown(plans: Array<{ batchDate: string; plan: HistoricalRecommendationPlan }>) {
  const closedPlans = [...plans]
    .filter(({ plan }) => isRecommendedPlan(plan) && isClosedResult(plan.outcome.result))
    .sort((left, right) => left.batchDate.localeCompare(right.batchDate));

  if (!closedPlans.length) {
    return null;
  }

  let equity = 100;
  let peak = 100;
  let worstDrawdown = 0;

  for (const { plan } of closedPlans) {
    equity *= 1 + plan.outcome.returnPct / 100;
    peak = Math.max(peak, equity);
    worstDrawdown = Math.max(worstDrawdown, ((peak - equity) / peak) * 100);
  }

  return roundMetric(worstDrawdown);
}

function confidenceBucketLabel(plan: HistoricalRecommendationPlan, horizon: HorizonId) {
  const threshold = HORIZON_THRESHOLDS[horizon];
  const score = plan.score ?? 0;

  if (score >= threshold + 8) {
    return "High confidence";
  }

  if (score >= threshold) {
    return "Tradable";
  }

  if (score >= threshold - 8) {
    return "Needs confirmation";
  }

  return "Low score";
}

function buildConfidenceCalibration(
  plans: HistoricalRecommendationPlan[],
  horizon: HorizonId
): ConfidenceCalibrationBucket[] {
  const buckets = new Map<string, HistoricalRecommendationPlan[]>();

  for (const plan of plans.filter(isRecommendedPlan)) {
    const label = confidenceBucketLabel(plan, horizon);
    buckets.set(label, [...(buckets.get(label) ?? []), plan]);
  }

  return ["High confidence", "Tradable", "Needs confirmation", "Low score"]
    .map((label) => {
      const bucketPlans = buckets.get(label) ?? [];
      const closedPlans = bucketPlans.filter((plan) => isClosedResult(plan.outcome.result));
      const wins = closedPlans.filter((plan) => isSuccessfulResult(plan.outcome.result)).length;

      return {
        label,
        total: bucketPlans.length,
        closed: closedPlans.length,
        averageScore: average(bucketPlans.map((plan) => plan.score ?? null)),
        hitRate: closedPlans.length ? roundMetric((wins / closedPlans.length) * 100) : null,
        averageReturnPct: average(closedPlans.map((plan) => plan.outcome.returnPct))
      };
    })
    .filter((bucket) => bucket.total > 0);
}

export function buildDailyPerformance(
  history: RecommendationDataset["history"],
  horizon: HorizonId
): DailyPerformance[] {
  return [...history]
    .sort((left, right) => right.batchDate.localeCompare(left.batchDate))
    .map((batch) => {
      const plans = batch.recommendations
        .map((recommendation) => recommendation.profiles[horizon])
        .filter((plan) => isRecommendedPlan(plan));
      const successful = plans.filter((plan) => isSuccessfulResult(plan.outcome.result)).length;
      const failed = plans.filter((plan) => plan.outcome.result === "stop_loss_hit").length;
      const open = plans.filter((plan) => plan.outcome.result === "open").length;
      const closed = successful + failed;

      return {
        batchDate: batch.batchDate,
        publishedAt: batch.publishedAt,
        total: plans.length,
        closed,
        open,
        successful,
        failed,
        successRate: closed ? roundMetric((successful / closed) * 100) : null,
        averageReturnPct: planAverageReturn(plans)
      };
    });
}

export function buildPerformanceSummary(rows: DailyPerformance[]) {
  const completedRows = rows.filter((row) => row.successRate !== null);
  const latestCompleted = completedRows[0] ?? null;

  if (!completedRows.length) {
    return {
      latestCompleted,
      averageSuccessRate: null,
      averageClosedReturnPct: null
    };
  }

  const averageSuccessRate =
    completedRows.reduce((total, row) => total + (row.successRate ?? 0), 0) / completedRows.length;

  const returnRows = completedRows.filter((row) => row.averageReturnPct !== null);
  const averageClosedReturnPct = returnRows.length
    ? returnRows.reduce((total, row) => total + (row.averageReturnPct ?? 0), 0) / returnRows.length
    : null;

  return {
    latestCompleted,
    averageSuccessRate: roundMetric(averageSuccessRate),
    averageClosedReturnPct:
      averageClosedReturnPct === null ? null : roundMetric(averageClosedReturnPct)
  };
}

export function buildBacktestEvaluation(dataset: RecommendationDataset): BacktestEvaluation {
  const horizons = dataset.profiles.map((profile) => {
    const datedPlans = dataset.history.flatMap((batch) =>
      batch.recommendations.map((recommendation) => ({
        batchDate: batch.batchDate,
        plan: recommendation.profiles[profile.id]
      }))
    );
    const publishedPlans = datedPlans
      .map(({ plan }) => plan)
      .filter(isRecommendedPlan);
    const closedPlans = publishedPlans.filter((plan) => isClosedResult(plan.outcome.result));
    const successful = closedPlans.filter((plan) => isSuccessfulResult(plan.outcome.result)).length;
    const benchmarkReturns = closedPlans
      .map((plan) => plan.outcome.benchmarkReturnPct)
      .filter((value): value is number => value !== null && value !== undefined && Number.isFinite(value));
    const benchmarkReturnPct = benchmarkReturns.length ? average(benchmarkReturns) : null;
    const averageReturnPct = average(closedPlans.map((plan) => plan.outcome.returnPct));

    return {
      horizon: profile.id,
      label: profile.label || HORIZON_LABELS[profile.id],
      total: publishedPlans.length,
      closed: closedPlans.length,
      open: publishedPlans.length - closedPlans.length,
      hitRate: closedPlans.length ? roundMetric((successful / closedPlans.length) * 100) : null,
      averageReturnPct,
      maxDrawdownPct: maxDrawdown(datedPlans),
      averageHoldingDays: average(closedPlans.map((plan) => plan.outcome.holdingDays)),
      benchmarkReturnPct,
      benchmarkCoverage: benchmarkReturns.length,
      alphaPct:
        averageReturnPct !== null && benchmarkReturnPct !== null
          ? roundMetric(averageReturnPct - benchmarkReturnPct)
          : null,
      confidenceCalibration: buildConfidenceCalibration(publishedPlans, profile.id)
    };
  });

  return {
    generatedAt: dataset.currentBatch.generatedAt,
    batchCount: dataset.history.length,
    benchmarkLabel: "Nifty 50 where generated history includes benchmark returns",
    horizons
  };
}

export function buildStockPerformanceHistory(
  history: HistoricalBatch[],
  symbol: string,
  horizon: HorizonId
): StockPerformanceHistoryEntry[] {
  return [...history]
    .sort((left, right) => right.batchDate.localeCompare(left.batchDate))
    .flatMap((batch) =>
      batch.recommendations
        .filter((recommendation) => recommendation.symbol === symbol)
        .filter((recommendation) => isRecommendedPlan(recommendation.profiles[horizon]))
        .map((recommendation) => ({
          batchDate: batch.batchDate,
          publishedAt: batch.publishedAt,
          symbol: recommendation.symbol,
          companyName: recommendation.companyName,
          sector: recommendation.sector,
          horizon,
          conviction: recommendation.profiles[horizon].conviction,
          entryPrice: recommendation.profiles[horizon].entryPrice,
          targetPrice: recommendation.profiles[horizon].targetPrice,
          stopLoss: recommendation.profiles[horizon].stopLoss,
          summary: recommendation.profiles[horizon].summary,
          outcome: recommendation.profiles[horizon].outcome
        }))
    );
}

export function outcomeLabel(result: OutcomeResult) {
  switch (result) {
    case "target_hit":
      return "Target hit";
    case "stop_loss_hit":
      return "Stop-loss hit";
    default:
      return "Open";
  }
}

export function outcomeClass(result: OutcomeResult) {
  switch (result) {
    case "target_hit":
      return "success";
    case "stop_loss_hit":
      return "danger";
    default:
      return "neutral";
  }
}

File: services/api/app/analytics.py
from .schemas import (
    BacktestEvaluation,
    ConfidenceCalibrationBucket,
    DailyPerformance,
    HistoricalBatch,
    HistoricalRecommendationPlan,
    HorizonBacktestSummary,
    HorizonId,
    OutcomeResult,
    RecommendationDataset,
    StockPerformanceHistoryEntry,
)

CLOSED_RESULTS: set[OutcomeResult] = {"target_hit", "stop_loss_hit"}
HORIZON_LABELS: dict[HorizonId, str] = {
    "single_day": "Single-day",
    "swing": "Swing",
    "position": "Position",
    "long_term": "Long-term",
}
HORIZON_THRESHOLDS: dict[HorizonId, float] = {
    "single_day": 54,
    "swing": 56,
    "position": 58,
    "long_term": 60,
}


def _round_metric(value: float | None) -> float | None:
    if value is None:
        return None

    return round(value, 2)


def _is_closed_result(result: OutcomeResult) -> bool:
    return result in CLOSED_RESULTS


def _is_recommended(plan: object) -> bool:
    return getattr(plan, "isRecommended", True)


def _average(values: list[float | int | None]) -> float | None:
    resolved = [float(value) for value in values if value is not None]

    if not resolved:
        return None

    return _round_metric(sum(resolved) / len(resolved))


def _max_drawdown(
    dated_plans: list[tuple[str, HistoricalRecommendationPlan]]
) -> float | None:
    closed_plans = sorted(
        [
            (batch_date, plan)
            for batch_date, plan in dated_plans
            if _is_recommended(plan) and _is_closed_result(plan.outcome.result)
        ],
        key=lambda item: item[0],
    )

    if not closed_plans:
        return None

    equity = 100.0
    peak = 100.0
    worst_drawdown = 0.0

    for _, plan in closed_plans:
        equity *= 1 + plan.outcome.returnPct / 100
        peak = max(peak, equity)
        worst_drawdown = max(worst_drawdown, ((peak - equity) / peak) * 100)

    return _round_metric(worst_drawdown)


def _confidence_bucket_label(plan: HistoricalRecommendationPlan, horizon: HorizonId) -> str:
    threshold = HORIZON_THRESHOLDS[horizon]
    score = plan.score or 0

    if score >= threshold + 8:
        return "High confidence"

    if score >= threshold:
        return "Tradable"

    if score >= threshold - 8:
        return "Needs confirmation"

    return "Low score"


def _build_confidence_calibration(
    plans: list[HistoricalRecommendationPlan], horizon: HorizonId
) -> list[ConfidenceCalibrationBucket]:
    buckets: dict[str, list[HistoricalRecommendationPlan]] = {}

    for plan in [item for item in plans if _is_recommended(item)]:
        label = _confidence_bucket_label(plan, horizon)
        buckets.setdefault(label, []).append(plan)

    rows: list[ConfidenceCalibrationBucket] = []

    for label in ["High confidence", "Tradable", "Needs confirmation", "Low score"]:
        bucket_plans = buckets.get(label, [])
        if not bucket_plans:
            continue

        closed_plans = [
            plan for plan in bucket_plans if _is_closed_result(plan.outcome.result)
        ]
        wins = sum(1 for plan in closed_plans if plan.outcome.result == "target_hit")
        rows.append(
            ConfidenceCalibrationBucket(
                label=label,
                total=len(bucket_plans),
                closed=len(closed_plans),
                averageScore=_average([plan.score for plan in bucket_plans]),
                hitRate=(
                    _round_metric((wins / len(closed_plans)) * 100)
                    if closed_plans
                    else None
                ),
                averageReturnPct=_average([plan.outcome.returnPct for plan in closed_plans]),
            )
        )

    return rows


def build_daily_performance(dataset: RecommendationDataset, horizon: HorizonId) -> list[DailyPerformance]:
    rows: list[DailyPerformance] = []

    for batch in sorted(dataset.history, key=lambda item: item.batchDate, reverse=True):
        plans = [
            recommendation.profiles[horizon]
            for recommendation in batch.recommendations
            if _is_recommended(recommendation.profiles[horizon])
        ]
        successful = sum(1 for plan in plans if plan.outcome.result == "target_hit")
        failed = sum(1 for plan in plans if plan.outcome.result == "stop_loss_hit")
        open_count = sum(1 for plan in plans if plan.outcome.result == "open")
        closed = successful + failed
        closed_returns = [
            plan.outcome.returnPct for plan in plans if _is_closed_result(plan.outcome.result)
        ]

        rows.append(
            DailyPerformance(
                batchDate=batch.batchDate,
                publishedAt=batch.publishedAt,
                total=len(plans),
                closed=closed,
                open=open_count,
                successful=successful,
                failed=failed,
                successRate=_round_metric((successful / closed) * 100) if closed else None,
                averageReturnPct=(
                    _round_metric(sum(closed_returns) / len(closed_returns))
                    if closed_returns
                    else None
                ),
            )
        )

    return rows


def build_stock_history(
    dataset: RecommendationDataset, symbol: str, horizon: HorizonId
) -> list[StockPerformanceHistoryEntry]:
    rows: list[StockPerformanceHistoryEntry] = []
    requested_symbol = symbol.upper()

    for batch in sorted(dataset.history, key=lambda item: item.batchDate, reverse=True):
        for recommendation in batch.recommendations:
            if recommendation.symbol != requested_symbol:
                continue

            plan = recommendation.profiles[horizon]
            if not _is_recommended(plan):
                continue
            rows.append(
                StockPerformanceHistoryEntry(
                    batchDate=batch.batchDate,
                    publishedAt=batch.publishedAt,
                    symbol=recommendation.symbol,
                    companyName=recommendation.companyName,
                    sector=recommendation.sector,
                    horizon=horizon,
                    conviction=plan.conviction,
                    entryPrice=plan.entryPrice,
                    targetPrice=plan.targetPrice,
                    stopLoss=plan.stopLoss,
                    summary=plan.summary,
                    outcome=plan.outcome,
                )
            )

    return rows


def build_backtest_evaluation(dataset: RecommendationDataset) -> BacktestEvaluation:
    horizons: list[HorizonBacktestSummary] = []

    for profile in dataset.profiles:
        dated_plans: list[tuple[str, HistoricalRecommendationPlan]] = [
            (batch.batchDate, recommendation.profiles[profile.id])
            for batch in dataset.history
            for recommendation in batch.recommendations
        ]
        published_plans = [plan for _, plan in dated_plans if _is_recommended(plan)]
        closed_plans = [
            plan for plan in published_plans if _is_closed_result(plan.outcome.result)
        ]
        successful = sum(1 for plan in closed_plans if plan.outcome.result == "target_hit")
        benchmark_returns = [
            plan.outcome.benchmarkReturnPct
            for plan in closed_plans
            if plan.outcome.benchmarkReturnPct is not None
        ]
        average_return = _average([plan.outcome.returnPct for plan in closed_plans])
        benchmark_return = _average(benchmark_returns)

        horizons.append(
            HorizonBacktestSummary(
                horizon=profile.id,
                label=profile.label or HORIZON_LABELS[profile.id],
                total=len(published_plans),
                closed=len(closed_plans),
                open=len(published_plans) - len(closed_plans),
                hitRate=(
                    _round_metric((successful / len(closed_plans)) * 100)
                    if closed_plans
                    else None
                ),
                averageReturnPct=average_return,
                maxDrawdownPct=_max_drawdown(dated_plans),
                averageHoldingDays=_average([plan.outcome.holdingDays for plan in closed_plans]),
                benchmarkReturnPct=benchmark_return,
                benchmarkCoverage=len(benchmark_returns),
                alphaPct=(
                    _round_metric(average_return - benchmark_return)
                    if average_return is not None and benchmark_return is not None
                    else None
                ),
                confidenceCalibration=_build_confidence_calibration(published_plans, profile.id),
            )
        )

    return BacktestEvaluation(
        generatedAt=dataset.currentBatch.generatedAt,
        batchCount=len(dataset.history),
        benchmarkLabel="Nifty 50 where generated history includes benchmark returns",
        horizons=horizons,
    )
